[FLINK-32132][table-planner] Cast function CODEGEN does not work as e…

zhougit86 opened the pull request:
## What is the purpose of the change

fix the default behaviour of cast exceptoin


## Brief change log

when the exception is thrown in legacy cast rule, the null variable is another name, we can return a null when exception thrown


## Verifying this change



This change is a trivial rework / code cleanup without any test coverage.




## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: ( no )
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: ( no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

zhougit86 commented:
> Can you please add a reproducible test to be sure that the PR fixes what it is expected to fix?

Thx for replay, I added a test: CastFunctionMiscITLegacyCase. if without the code I added, the test could return a value -1, which is not expected

zhougit86 commented:
@flinkbot run azure

zhougit86 commented:
@flinkbot run azure

zhougit86 commented:
@snuyanzin Hi Master, Could you please help review the test case and let me know if it makes sense....

luoyuxia commented:
@zhougit86 Thanks for contribution. But i'm still trying to understand what's the problem and why your changes can fix it. Could you please provide a reproducible case so that I can try to reproduce it in my local and to understand the problem.

zhougit86 commented:
@luoyuxia please find the test case in the commit, when the "cast( a as bigint)" function's input type is "STRING().notNull()", we suppose to get a null value. but now it returns a default value "-1". this is not correct.

In the test case, I made it throw a not null enforcer exception. without the change, I will return a -1, which I think is not correct.

snuyanzin commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionMiscITLegacyCase.java:
```diff
@@ -0,0 +1,60 @@
+    @Override
+    Stream<TestSetSpec> getTestSetSpecs() {
+        return Stream.of(
+                TestSetSpec.forFunction(BuiltInFunctionDefinitions.CAST, "try cast throw exception")
```
this description is confusing: are we testing `CAST` or `TRY_CAST`?

snuyanzin commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala:
```diff
@@ -556,7 +556,7 @@ object CodeGenUtils {
       }
 
-      if (fieldType.isNullable) {
+      if (fieldType.isNullable || fieldExpr.nullTerm.startsWith("castRuleResultIsNull")) {
```
IMHO it is not good to rely on var namings
rather it's better to check properties
why can't we pass a corresponding arg to `org.apache.flink.table.planner.codegen.calls.ScalarOperatorGens$#generateCast` for `nullOnFailure` ?

zhougit86 commented:
> @zhougit86 Thanks for contribution. But i'm still trying to understand what's the problem and why your changes can fix it. Could you please provide a reproducible case so that I can try to reproduce it in my local and to understand the problem.

@luoyuxia could you please help review? thx

snuyanzin commented:
hi @zhougit86 
it looks ok from my side thanks for addressing comments

the only nit pick: could you please rename `CastFunctionMiscITLegacyCase` in way that it matches regex `'[A-Z][A-Za-z\d]*Test(s|Case)?|Test[A-Z][A-Za-z\d]*|IT(.*)|(.*)IT(Case)`
e.g. something like `CastFunctionMiscLegacyITCase`

zhougit86 commented:
> hi @zhougit86 it looks ok from my side thanks for addressing comments
> 
> the only nit pick: could you please rename `CastFunctionMiscITLegacyCase` in way that it matches regex `'[A-Z][A-Za-z\d]*Test(s|Case)?|Test[A-Z][A-Za-z\d]*|IT(.*)|(.*)IT(Case)` e.g. something like `CastFunctionMiscLegacyITCase`

Thx for the review master, fixed @snuyanzin

snuyanzin commented:
hmm, it looks like you renamed class however not renamed file... could you also please rename file in order to have them in sync?

snuyanzin commented:
also could you please rebase to the latest master?
The failure was fixed in later commits, so rebase should help with passing tests

zhougit86 commented:
> also could you please rebase to the latest master? The failure was fixed in later commits, so rebase should help with passing tests

thx for the instruction... my bad not modify file name .... fixed @snuyanzin

snuyanzin commented:
@flinkbot run azure

zhougit86 commented:
> ## CI report:
> * [9b777cc](https://github.com/apache/flink/commit/9b777cc410c0075201b7d24158c6cfe2aa368b15) Azure: [SUCCESS](https://dev.azure.com/apache-flink/98463496-1af2-4620-8eab-a2ecc1a2e6fe/_build/results?buildId=50286)
> 
> Bot commands

@snuyanzin Master, not sure whether the rebase took effect? :)

snuyanzin reviewed (approved):
LGTM
thanks, will merge soon

snuyanzin commented:
@zhougit86 sorry for the late response

I have a question: could you please clarify it
it seems your change is inside condition
```scala
if (codeGeneratorCastRule.canFail(inputType, targetType) && nullOnFailure) {
...
val adjustedTargetType = if (nullOnFailure) targetType.copy(nullOnFailure) else targetType
          return GeneratedExpression(
            resultTerm,
            nullTerm,
            operand.code + "\n" + castCode,
            adjustedTargetType
          )
...
} else {
...
          return GeneratedExpression(
            castCodeBlock.getReturnTerm,
            castCodeBlock.getIsNullTerm,
            operand.code + castCode,
            targetType
          )
}
```
shouldn't we extract `val adjustedTargetType = if (nullOnFailure) targetType.copy(nullOnFailure) else targetType` from `if/else` and apply it to both `if` and `else` branches?

zhougit86 commented:
> @zhougit86 sorry for the late response
> 
> I have a question: could you please clarify it it seems your change is inside condition
> 
> ```scala
> if (codeGeneratorCastRule.canFail(inputType, targetType) && nullOnFailure) {
> ...
> val adjustedTargetType = if (nullOnFailure) targetType.copy(nullOnFailure) else targetType
>           return GeneratedExpression(
>             resultTerm,
>             nullTerm,
>             operand.code + "\n" + castCode,
>             adjustedTargetType
>           )
> ...
> } else {
> ...
>           return GeneratedExpression(
>             castCodeBlock.getReturnTerm,
>             castCodeBlock.getIsNullTerm,
>             operand.code + castCode,
>             targetType
>           )
> }
> ```
> 
> shouldn't we extract `val adjustedTargetType = if (nullOnFailure) targetType.copy(nullOnFailure) else targetType` from `if/else` and apply it to both `if` and `else` branches?

Hate to admit my negligence being caught again :(

I think maybe we shouldn't extract ```val adjustedTargetType = if (nullOnFailure) targetType.copy(nullOnFailure) else targetType```  out from the ```if/else``` section, as in the else section, we will not catch the Exception thrown by the cast, so the nullTerm is actually useless in the generated code. 

Anyway, I will remove the redundant ```if (nullOnFailure)```, let me know your think please...

zhougit86 commented:
@flinkbot run azure

snuyanzin commented:
thanks for addressing comment
great that it's becoming even smaller

snuyanzin commented:
@zhougit86 thanks for the contribution
are you willing to create backports to 1.17.x, 1.16.x?

zhougit86 commented:
> 

sure thing, will do quickly

zhougit86 commented:
https://github.com/apache/flink/pull/23028
https://github.com/apache/flink/pull/23029
@snuyanzin Hi master, not sure if this is what you mean create backports?
